XOAuth2 implementation for Google Accounts

philipwhiuk opened the pull request:
This is probably now in a reasonable position for review. There's still some extended testing to do as I mentioned in the issue. But there's (unless recommended from review) unlikely to be any huge refactoring, just some tweaks round the edges. 

Fixes #655

philipwhiuk commented:
Recent work has focused on token refresh. Hopefully unless you manually de-auth the app it should just work in any scenario.

Next I need to work out what happens if you remove a Google account. 

I happen to have a Google Apps e-mail. So I plan to use that for initial testing. If I run into problems with that I'll split it out and just use a regular Gmail account and make a note of the issues.

Valodim commented:
Awesome work! :+1:

dschuermann commented:
Great effort, I really like to see this XOAuth2 support baked into K-9 Mail in general independent of GMail, a much better SASL authentication technique.

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapConnection.java:
```diff
@@ -356,6 +371,46 @@ private void authenticate() throws MessagingException, IOException {
+    }
+
+    private void attemptXOAuth2() throws MessagingException, IOException {
+        String command = Commands.AUTHENTICATE_XOAUTH2;
```
putting this in a variable seems unnecessary, maybe put the `getToken` output in a variable instead to make the sendSaslIrCommand call more clear?

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapConnection.java:
```diff
@@ -635,6 +690,31 @@ protected String getLogId() {
+                if (sensitive && !K9MailLib.isDebugSensitive()) {
+                    Log.v(LOG_TAG, getLogId() + ">>> [Command Hidden, Enable Sensitive Debug Logging To Show]");
+                } else {
+            Log.v(LOG_TAG,  getLogId() + ">>> " + tag + " " + command+ " " + initialClientResponse);
```
indentation

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/RemoteStore.java:
```diff
@@ -51,7 +54,10 @@ public synchronized static Store getInstance(Context context,
-                        (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE));
+                        (ConnectivityManager) context.getSystemService(
+                                Context.CONNECTIVITY_SERVICE),
+                        oAuth2TokenProvider
```
indentation

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/transport/SmtpTransport.java:
```diff
@@ -300,18 +305,22 @@ public void open() throws MessagingException {
             }
             parseOptionalSizeValue(extensions);
 
             if (mUsername != null
```
`TextUtils.isEmpty`

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/transport/SmtpTransport.java:
```diff
@@ -757,6 +772,45 @@ private void saslAuthCramMD5(String username, String password) throws MessagingE
+        try {
+            attemptXoauth2(username);
+        } catch (NegativeSmtpReplyException exception) {
+            if (exception.getReplyCode() == 535) {
```
magic constant

Valodim commented:
```
  1562   GoogleSignatureVer..  V  com.fsck.k9.debug signature not valid.  Found:
  1562   GoogleSignatureVer..  V  MIIDDTCCAfWgAwIBAgIEcNtjeDANBgkqhkiG9w0BAQsFADA3MQswCQYDVQQGEwJVUzEQMA4GA1UE
...
```

```
GLSActivity  W  ehh: NeedPermission
GLSActivity  W  at egu.a(SourceFile:1526)
GLSActivity  W  at egt.a(SourceFile:920)
GLSActivity  W  at egt.e(SourceFile:534)
GLSActivity  W  at egs.a(SourceFile:37)
GLSActivity  W  at ehv.getAuthToken(SourceFile:179)
GLSActivity  W  at android.accounts.AbstractAccountAuthenticator$Transport.getAuthToken(AbstractAccountAuthenticator.java:196)
GLSActivity  W  at android.accounts.IAccountAuthenticator$Stub.onTransact(IAccountAuthenticator.java:113)
GLSActivity  W  at android.os.Binder.execTransact(Binder.java:446)
         k9  E  Failed to login, closing connection for conn69127108
         k9  E  Error while testing settings (auth failed)
         k9  E  com.fsck.k9.mail.AuthenticationFailedException: Attempt to invoke virtual method 'boolean java.lang.Object.equals(java.lang.Object)' on a null object reference
         k9  E  at com.fsck.k9.account.AndroidAccountOAuth2TokenStore.fetchNewAuthToken(AndroidAccountOAuth2TokenStore.java:109)
         k9  E  at com.fsck.k9.account.AndroidAccountOAuth2TokenStore.getToken(AndroidAccountOAuth2TokenStore.java:82)
         k9  E  at com.fsck.k9.mail.store.imap.ImapConnection.attemptXOAuth2(ImapConnection.java:399)
         k9  E  at com.fsck.k9.mail.store.imap.ImapConnection.authXoauth2withSASLIR(ImapConnection.java:376)
         k9  E  at com.fsck.k9.mail.store.imap.ImapConnection.authenticate(ImapConnection.java:335)
         k9  E  at com.fsck.k9.mail.store.imap.ImapConnection.open(ImapConnection.java:129)
         k9  E  at com.fsck.k9.mail.store.imap.ImapStore.checkSettings(ImapStore.java:308)
         k9  E  at com.fsck.k9.activity.setup.AccountSetupCheckSettings$CheckAccountTask.checkIncoming(AccountSetupCheckSettings.java:496)
         k9  E  at com.fsck.k9.activity.setup.AccountSetupCheckSettings$CheckAccountTask.checkServerSettings(AccountSetupCheckSettings.java:465)
         k9  E  at com.fsck.k9.activity.setup.AccountSetupCheckSettings$CheckAccountTask.doInBackground(AccountSetupCheckSettings.java:422)
         k9  E  at com.fsck.k9.activity.setup.AccountSetupCheckSettings$CheckAccountTask.doInBackground(AccountSetupCheckSettings.java:400)
         k9  E  at android.os.AsyncTask$2.call(AsyncTask.java:292)
         k9  E  at java.util.concurrent.FutureTask.run(FutureTask.java:237)
         k9  E  at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:231)
         k9  E  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
         k9  E  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
         k9  E  at java.lang.Thread.run(Thread.java:818)
```

Valodim commented on k9mail/src/main/java/com/fsck/k9/account/AndroidAccountOAuth2TokenStore.java:
```diff
@@ -0,0 +1,127 @@
+                             final OAuth2TokenProviderAuthCallback callback) {
+        Account account = getAccountFromManager(emailAddress);
+        if (account == null) {
+            callback.failure(new Exception(activity.getString(R.string.xoauth2_account_doesnt_exist)));
```
raw `Exception` type? hmm

Valodim commented on k9mail/src/main/java/com/fsck/k9/activity/setup/AccountSetupBasics.java:
```diff
@@ -64,10 +67,13 @@
     private EditText mPasswordView;
     private CheckBox mClientCertificateCheckBox;
     private ClientCertificateSpinner mClientCertificateSpinner;
+    private CheckBox mOAuth2CheckBox;
```
codestyle, no more m prefix

philipwhiuk replied:
It looks weird if only one variable is different.

Valodim commented on k9mail/src/main/java/com/fsck/k9/activity/setup/AccountSetupOutgoing.java:
```diff
@@ -410,11 +417,16 @@ private void validateFields() {
                         && Utility.requiredFieldValid(mPortView)
                         && (!mRequireLoginView.isChecked()
-                                || hasValidPasswordSettings || hasValidExternalAuthSettings));
+                                || hasValidPasswordSettings
```
codestyle, operators at the end of the previous line

Valodim commented on k9mail/src/main/java/com/fsck/k9/activity/setup/AccountSetupIncoming.java:
```diff
@@ -538,55 +554,96 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
-            ConnectionSecurity connectionSecurity = getSelectedSecurity();
+        AuthType authType = getSelectedAuthType();
+        if (authType == AuthType.XOAUTH2) {
+            K9.oAuth2TokenStore.authorizeAPI(mAccount.getEmail(), this,
```
this block probably deserves to be its own method

philipwhiuk replied:
But it makes it less clear that you're waiting for the API authorization before continuing (the success block takes the place of the stuff done if XOAUTH isn't being used.

Valodim commented on k9mail/src/main/java/com/fsck/k9/activity/setup/AccountSetupIncoming.java:
```diff
@@ -538,55 +554,96 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
-                extra.put(WebDavStoreSettings.MAILBOX_PATH_KEY,
-                        mWebdavMailboxPathView.getText().toString());
+    private void fetchAuthTokenForAccount(final String emailAddress) {
+        new Exception().printStackTrace();
```
debugging relic

philipwhiuk replied:
The whole method actually is now not needed!

Valodim commented on k9mail/src/main/java/com/fsck/k9/activity/setup/AccountSetupIncoming.java:
```diff
@@ -538,55 +554,96 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
+        android.accounts.Account[] accounts = accountManager.getAccountsByType("com.google");
+        for (android.accounts.Account account : accounts) {
+            Log.w(K9.LOG_TAG, "Account: " + account.name);
+            if(account.name.equals(emailAddress)) {
```
codestyle, missing space

Valodim commented on k9mail/src/main/java/com/fsck/k9/activity/setup/AccountSetupIncoming.java:
```diff
@@ -538,55 +554,96 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
+                            public void run(AccountManagerFuture<Bundle> future) {
+                                try {
+                                    Bundle bundle = future.getResult();
+                                    if(bundle.get(AccountManager.KEY_ACCOUNT_NAME).equals(emailAddress)) {
```
empty block?

Valodim commented:
I believe the ui flow for this feature should be different. At the moment it's "Advanced options -> Use OAuth 2.0 token", when what this really does is "If I use gmail, make everything magically work". Given the huge use case that is gmail, this workflow should surely be displayed in a more prominent way

Valodim commented:
![])(https://dl2.pushbulletusercontent.com/Zm1jGunJrIWBRjHl4QOXRat66VpWhtcG/Screenshot_2016-05-21-11-54-15.png)

philipwhiuk commented:
Thanks for the feedback. I'll work on it this weekend.

philipwhiuk commented:
I've rebased on top of master and fixed up various bits in response to the review. I've made the case where you got the exception behave better but I'm not quite sure why it occurred. It's possible I'm still missing some failure case handling.

I'm still missing the API key from @cketti - right now you have to approve an arbitrary app which I can't find how to revoke. So it's not really acceptable to merge right now.

philipwhiuk commented:
I've finally got back around to this and re-based on to current master. Still need the key to finish it off though :(

dschuermann commented:
I said it before, but I consider this an important feature for K-9 Mail adoption for people switching from the GMail app to K-9 Mail.

Valodim commented on k9mail/src/main/java/com/fsck/k9/account/AndroidAccountOAuth2TokenStore.java:
```diff
@@ -0,0 +1,136 @@
+            accountManager.getAuthToken(account, GMAIL_AUTH_TOKEN_TYPE, null, activity,
+                new AccountManagerCallback<Bundle>() {
+                    @Override
+                    public void run(AccountManagerFuture<Bundle> future) {
```
this method has borderline complexity for its nesting depth.. given its if/else/ends-in-callback structure though I don't think the code would be more readable otherwise, so might as well leave it as is

Valodim commented on k9mail/src/main/java/com/fsck/k9/K9.java:
```diff
@@ -506,6 +508,7 @@ public void onCreate() {
         super.onCreate();
         app = this;
         Globals.setContext(this);
+        oAuth2TokenStore = new AndroidAccountOAuth2TokenStore(this);
```
we have the `Globals` class for this now :tada:

Valodim commented on k9mail-library/src/test/java/com/fsck/k9/mail/store/imap/ImapConnectionTest.java:
```diff
@@ -275,6 +282,105 @@ public void open_authCramMd5WithoutAuthCramMd5Capability_shouldThrow() throws Ex
     }
 
     @Test
+    public void open_authXoauthWithSaslIr() throws Exception {
```
please stick with the pattern the other methods in this test class use, it's very difficult to see where setup, test and assertions are otherwise

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/transport/SmtpTransport.java:
```diff
@@ -300,18 +308,21 @@ public void open() throws MessagingException {
-            if (mUsername != null
-                    && mUsername.length() > 0
-                    && (mPassword != null && mPassword.length() > 0 || AuthType.EXTERNAL == mAuthType)) {
+            if (TextUtils.isEmpty(mUsername)
```
`!= null` becomes `TextUtils.isEmpty` here? is this right?

philipwhiuk replied:
```
mUsername != null && mUsername.length() > 0
```

is equivalent to just using TextUtils.isEmpty. So this is cleaner.

philipwhiuk replied:
Oops. Looks like there's no tests at all on `SmtpTransport.open()`

I think I'll have a crack at making that testable (and thus this would have created a regression I think) in a separate PR.

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapStoreUriDecoder.java:
```diff
@@ -83,7 +83,7 @@ public static ImapStoreSettings decode(String uri) {
             String[] userInfoParts = userinfo.split(":");
 
             if (userinfo.endsWith(":")) {
```
(not related to your changes, but wouldn't `userInfoParts.length < 2 || TextUtils.isEmpty(userInfoParts[1])` be a more reliable check here?)

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/Authentication.java:
```diff
@@ -82,4 +83,10 @@ public static String computeCramMd5(String username, String password, String b64
+    public static String computeXoauth(String username, String authToken) throws UnsupportedEncodingException {
+        return new String(Base64.encodeBase64(
+                ("user="+username+"\001auth=Bearer "+authToken+"\001\001").getBytes()
+        ),US_ASCII);
```
code style

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/Authentication.java:
```diff
@@ -82,4 +83,10 @@ public static String computeCramMd5(String username, String password, String b64
+
+    public static String computeXoauth(String username, String authToken) throws UnsupportedEncodingException {
+        return new String(Base64.encodeBase64(
+                ("user="+username+"\001auth=Bearer "+authToken+"\001\001").getBytes()
```
a `String.format` thing would be nicer to read here I think

philipwhiuk commented:
I noticed during testing (I have a Gmail account so can dogfood this fairly easily)  that some changes from code review had broken the URI behaviour. So I've changed that code around a bit and add a whole bunch of extra test cases for ImapStoreUri decoding.

I've also added tests for the SettingsImporter. This is an untested front-end class. So I wrote just enough to be able to test the bit of code I changed: https://github.com/k9mail/k-9/pull/1295/files#diff-46c0f57bfd70565c0aa5f0f703f4917eR131

There's lots of scope for expanding these tests to cover missing settings (e.g. no identity, no server field), duplicate settings, etc.

jberkel commented:
Nice to see XOAuth2 is happening. I already implemented basic support in my fork for another Android app I'm working on which uses the k9-lib (https://github.com/jberkel/k-9/commit/3113a130f77af2bba21ef9966f911cf07c3e3c53). 

Would be great to have your patch in mainline. Can I help in any way?

philipwhiuk commented:
I think @cketti plans to merge it after we do a stable release. There's already a lot in unstable waiting for stable release so we're trying to avoid large code changes at this point.

jberkel commented:
haven't followed the K9 release process in a while but having lots of unmerged/open PRs is usually not a good idea. a way around that would be to merge in the backend/server code first and activate the front-end at a later stage. makes the PRs also more manageable, in case they can be split nicely.

philipwhiuk commented:
Splitting apart sounds great in theory but it wouldn't actually give us anything useful in terms of real world testing.

This particular release is the PGP/MIME stuff so I think it's been unduly large and slow to close. And everything nextStable is built on top of that code (which changed a lot of the core so there was no way to avoid that). Once it drops, I broadly expect us to immediately merge the stuff in next stable and then push a new beta on the 5.3x branch fairly soon after.

jberkel commented:
Yes, real world testing not exactly, but the new xoauth2 code could be merged into the mainline without being activated. Once deployed this would make sure that there are no regressions in the other auth-related parts of the code. It would also avoid merge conflicts. 

Is there a rough release date for the new PGP stuff?

philipwhiuk commented:
I've just pushed some SMTP fixes and tests that were possible with the new MockSmtpServer I wrote. Basically I've introduced a new structure to handle CommandResponses better and properly dealt with the challenge response from Google. I've tested a lot of the cases XOAUTH2 can generate as a result.

The structure was desired anyway (there was a TODO to start looking at response codes more which I've shuffled to a method comment). I've also deprecated some methods - we should start migrating to the new version which allows you access to the response code.

jberkel commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapConnection.java:
```diff
@@ -356,6 +368,46 @@ private void authenticate() throws MessagingException, IOException {
+            //We could avoid this if we had a reasonable chance of knowing
+            //if a token was invalid before use (e.g. due to expiry). But we don't
+            //This is the intended behaviour per AccountManager
+            Log.v(LOG_TAG, "Authentication exception, invalidating token and re-trying", e);
```
Might make sense to decode the JSON error payload, it contains some more information about why the request failed:

https://developers.google.com/gmail/xoauth2_protocol

We don't need to retry in all cases.

philipwhiuk replied:
Thanks. I'll add in the decoding I just added to SMTP. I'm not necessarily sure the cases in which retry is the correct solution right now (or even all the cases).

philipwhiuk replied:
It's possible they are just HTTP Status codes - the SMTP example I added from Google's documentation has 401 (Unauthorised). Maybe I should not retry on 5xx error codes.

philipwhiuk commented:
Right, the latest commits covers parsing of those JSON responses both in SMTP and IMAP.

jberkel commented on k9mail-library/src/main/java/com/fsck/k9/mail/transport/SmtpTransport.java:
```diff
@@ -756,6 +829,71 @@ private void saslAuthCramMD5(String username, String password) throws MessagingE
+
+            oauthTokenProvider.invalidateToken(username);
+
+            handlePermanentFailure(negativeResponseFromNewToken);
```
this whole block duplicates a lot of of the previous method. maybe encapsulate the general approach, and see if it could also be used in `ImapConnection`.

jberkel commented on k9mail-library/src/test/java/com/fsck/k9/mail/XOAuth2ChallengeParserTest.java:
```diff
@@ -0,0 +1,15 @@
+    public static final String MISSING_STATUS_RESPONSE = Base64.encode(
+            "{\"schemes\":\"bearer mac\",\"scope\":\"https://mail.google.com/\"}");
+    public static final String INVALID_RESPONSE = Base64.encode(
+            "{\"status\":\"401\",\"schemes\":\"bearer mac\",\"scope\":\"https://mail.google.com/\"");
```
no tests?

philipwhiuk replied:
Essentially this class is tested by proxy from the Smtp and IMAP class. I think the coverage is done completely.

jberkel commented on k9mail-library/src/test/java/com/fsck/k9/mail/transport/SmtpTransportTest.java:
```diff
@@ -175,6 +189,205 @@ public void open_withoutCramMd5AuthExtension_shouldThrow() throws Exception {
+        server.expect("");
+        server.output("535-5.7.1 Username and Password not accepted. Learn more at");
+        server.output("535 5.7.1 http://support.google.com/mail/bin/answer.py?answer=14257 hx9sm5317360pbc.68");
+        server.expect("AUTH XOAUTH2 dXNlcj11c2VyAWF1dGg9QmVhcmVyIG5ld1Rva2VuAQE=");
```
shouldn't this be a different token?

philipwhiuk replied:
It's not JSON. JSON is only used as a response by Google. But I agree about more readable.

jberkel commented on k9mail-library/src/test/java/com/fsck/k9/mail/transport/SmtpTransportTest.java:
```diff
@@ -175,6 +189,205 @@ public void open_withoutCramMd5AuthExtension_shouldThrow() throws Exception {
+        server.output("250-localhost Hello client.localhost");
+        server.output("250 AUTH XOAUTH2");
+        server.expect("AUTH XOAUTH2 dXNlcj11c2VyAWF1dGg9QmVhcmVyIG9sZFRva2VuAQE=");
+        server.output("334 "+ XOAuth2ChallengeParserTest.INVALID_RESPONSE);
```
Maybe some of these tests could be done a more unit-test level, without requiring a server to be run? This would be faster and require less maintenance. Then use the server tests just for a few basic tests.

philipwhiuk commented on k9mail-library/src/main/java/com/fsck/k9/mail/transport/SmtpTransport.java:
```diff
@@ -756,6 +829,71 @@ private void saslAuthCramMD5(String username, String password) throws MessagingE
+        try {
+            attemptXoauth2(username);
+        } catch (NegativeSmtpReplyException negativeResponse) {
+            if (negativeResponse.getReplyCode() != SMTP_AUTHENTICATION_FAILURE_ERROR_CODE) {
```
I'll look at this (although it'll just involve catching the NegativeSmtpReplyException and throwing something that encapsulates it).

philipwhiuk commented:
I'm going to work this week on #1698 It'll be on top of this, but it'll probably modify the UX a bit so I'm hoping to get it in the same release to avoid confusion.

My current design idea is:

* Change the first screen to only take an email address
* If the email address is an Android Account email, try and automatically configure it to use XOAUTH.
* If the email address is a Yahoo email, try and automatically configure it to use XOAUTH with the Yahoo URL.
* If not, have password & certificate screen which continues the current system - i.e. if it's configured, use the autoconfiguration, otherwise go to the IMAP / etc selection screen.

The Yahoo XOAUTH URL will be stored in config.

The OAuthProvider will be told / know which to use.

cketti commented:
I'm sorry I've been largely ignoring this PR for the past few weeks. A couple of comments:
* This PR is already quite large. Adding even more to it makes it that much harder to review and get it merged.
* I plan to overhaul the whole UI. Especially the setup flow. So making changes now that don't fit into the "new concept" is probably a waste of time.
* Sadly, XOAUTH2 is not a documented standard that email clients can implement in a provider-independent way. I'm not sure we want to support XOAUTH2 with Yahoo. With their proprietary Push implementation and the recent revelations they don't exactly have a good track record.

My suggestion is to extract the "backend" part and associated tests into a new pull request. We should be able to merge that code right away without impacting stability because it wouldn't be used.
Then we work on a concept for a new setup experience and how we best fit the OAuth2 flow (and all relevant error cases) into it.

philipwhiuk commented:
That's fine. I'm waiting to hear back from Yahoo on how I get access to the Mail permission in their XOAUTH implementation anyway. I'll re-base this into two commits - a front-end one and a back-end one.

philipwhiuk commented:
Closing this as I'll open a new one for the front-end changes later.
